dot_marketplace-Phase3-milestone_1, 2 & 3 submission

Loaki07 opened the pull request:
# Milestone Delivery Checklist

- [x] The [milestone-delivery-template.md](https://github.com/w3f/Grant-Milestone-Delivery/blob/master/deliveries/milestone-delivery-template.md) has been copied and updated.
- [x] The [invoice form :pencil:](https://forms.gle/LSRr7PCjBpEbKGh89) has been filled out for this milestone.
- [x] This pull request is being made by the same account as the accepted application.
- [x] In case of acceptance, the payment will be transferred to the BTC/ETH payment address provided in the application.
- [x] The delivery is according to the [Guidelines for Milestone Deliverables](https://github.com/w3f/Grants-Program/blob/master/docs/Support%20Docs/milestone-deliverables-guidelines.md).

Link to the application pull request: https://github.com/w3f/Grants-Program/pull/1205

0xCaso commented on deliveries/dot_marketplace-Phase3-milestone_1.md:
```diff
@@ -0,0 +1,17 @@
+
+**The [invoice form :pencil:](https://docs.google.com/forms/d/e/1FAIpQLSdSqj2vYjvpiIytkjcc40Pwl0Eg76WGUAq5L9e8eFuuOegmLw/viewform) has been filled out correctly for this milestone and the delivery is according to the official [milestone delivery guidelines](https://github.com/w3f/General-Grants-Program/blob/master/grants/milestone-deliverables-guidelines.md).**
+
+* **PR Link:** https://github.com/w3f/Grants-Program/pull/1205
```
Hi @Loaki07, thanks a lot for the delivery, someone will look into it asap. 

In the meantime, could you do this little edit to fix our workflow?
```suggestion
* **Application Document:** [dot_marketplace-Phase3.md](https://github.com/w3f/Grants-Program/blob/master/applications/dot_marketplace-Phase3.md)
```

Loaki07 commented:
Update: Have submitted the dot_marketplace-Phase3-milestone_2 delivery as well.

Loaki07 commented:
Update: Submitted the final milestone dot_marketplace-Phase3-milestone_3 delivery as well.

0xCaso commented:
Hi @Loaki07, thanks for the 3 deliveries, and sorry for the waiting. I'll begin the evaluations asap, and I'll reach out if I need your help.

0xCaso reviewed (commented):
Here I am @Loaki07. Before accepting the first milestone, I'd like to write the following comments/questions:
- Have you considered/planned to make `TaskTypeTags` customizable/editable?
- Is the milestone's `deadline` the number of days from the _start_ of the project, or something else? In this case, shouldn't you do some checks like M1 deadline should be **<** than M2 deadline? Or do you want to let this free?
- Do you plan to control the `rating` value? For now it should be a number from 0 to 255, as now it's `u8`;
- In the videoguide, **Alice** manages to close the project also if the milestone `1b` was still open. As I couldn't, and I only could close it if all the milestones are completed, I assume that you've seen the issue and solved it, but it was still in the video, right? In this case it's not a problem, I just want to double check.

0xCaso commented:
Another thing: i was starting the evaluation of Milestone 2, and I've seen you delivered also a frontend in the following repo: https://github.com/WowLabz/tasking_frontend; here, I can see there is a branch named `Phase3_Milestone1`. Is that supposed to be evaluated, or one of the deliveries of Milestone 1?

0xCaso reviewed (commented):
# Milestone 2

As for the previous milestone, both docker and without-docker setups were straightforward.
<br/>

## Backend comments

Everything seems to work greatly, I've just two things to say about `votes` and `bids`:

- ### `accept_jury_duty` issue

  I've just noted an issue, but I'm not sure if it's on your end, so I'll ask here: often, when I try to cast a vote with a `Potential Juror` I get this error from the frontend:
  
  ![Screenshot 2023-01-12 at 13 27 24](https://user-images.githubusercontent.com/22419450/212080260-99e489d0-7dff-43fa-bff5-082cd2ac6afc.png)
  
  So I tried to reproduce it with PolkadotJS to know more about it, and it gave me the `JurySelectionPeriodElapsed` error. From the [source code](https://github.com/WowLabz/dot-marketplace-v2/blob/bbeb45c2dfeeb91d2a4e39d191658361375992ef/pallets/pallet-tasking/src/lib.rs#L920), I see that it's like the `Juror` has been too slow to fire the `accept_jury_duty` function. Is that possible, and is that a bug?

- ### `accept_jury_duty` ratings
  (This is just a discussion to understand better the project's operations)
  
  I didn't get it why a `Juror` can, as well as choosing who they are voting for, cast their ratings for both the `customer` and `worker`. Is this to "correct" the ratings they're voting against? If yes, is it good for milestones disputes, as there are no ratings at the time a milestone is disputed? Or maybe is because a voter can express his rating in advance?
<br/>

## Frontend comments

Other than this, I didn't see any other important things in the backend, but I'd like to comment quickly the frontend (I think it should be an MVP, but maybe these could still be useful to you): 
- On the `login` page, I’d replace `User Name` with `Email` (and not writing under the input field “Enter registered EmailID”)
- On the `create project` page, if I try to change the wallet the frontend breaks, and then I have to logout and re-login;
- If I try `edit milestone`, its cost increases each time the button gets pressed;
- Isn't it possible to add an image and upload it in just one step instead of two?
- It looks a bit weird that just the `customer`, initially, must signup and login, but I guess it's for the MVP; by the way, what's the purpose of the signup?
- Milestone's button is `Complete` for a `worker` who bidded also if the `customer` accepted a bid of another `worker` (if you then click it, it says "project does not exist").
<br/>

## Conclusion

To accept the milestone, i'd like to get answers for the **Backend** part, but the **Frontend** one is definetly not urgent, as those are just comments for improvement / bugs findings.

Loaki07 commented:
> # Milestone 1
> Here I am @Loaki07. Before accepting the first milestone, I'd like to write the following comments/questions:
> 
> * Is it ok that a user can bid the same milestone multiple times, also if his other bid is still open?
> * Have you considered/planned to make `TaskTypeTags` customizable/editable?
> * The milestone's `deadline` is the number of day the milestone can last, right? So in this case M2's deadline can be < than M1's;
> * Do you plan to control the `rating` value? For now it should be a number from 0 to 255, as now it's `u8`; **EDIT**: I've seen in milestone 2 this is checked (1<=rating<=5).
> * In the videoguide, **Alice** manages to close the project also if the milestone `1b` was still open. As I couldn't, and I only could close it if all the milestones are completed, I assume that you've seen the issue and solved it, but it was still in the video, right? In this case it's not a problem, I just want to double check.

1. We have fixed this issue for all the milestones. A bidder can now only bid once per milestone.
2. Yes, as per the clients we can modify this to include tags of their own. However, we can can also modify this to be dynamic as per request by introducing a Tag Management Model.
3. Yes, the milestones are independent of each others timelines (based on their complexity).
4. Yes, its fixed to a number between 1-5 both in the frontend and backend.
5. Yes, this was possible because all the milestone was still `Open`. In this case, the funds to all the bidders will be transferred back [code](https://github.com/WowLabz/dot-marketplace-v2/blob/Phase3_Milestone1/pallets/pallet-tasking/src/lib.rs#L1083-L1116). However, when each milestone in the project is not in either status `Open` or `Completed` the customer can't close the project. This was designed to be this way to give the customer an option to close a project that he might have added and doesn't want to proceed with it.

Loaki07 commented:
> Another thing: i was starting the evaluation of Milestone 2, and I've seen you delivered also a frontend in the following repo: https://github.com/WowLabz/tasking_frontend; here, I can see there is a branch named `Phase3_Milestone1`. Is that supposed to be evaluated, or one of the deliveries of Milestone 1?
> 
This branch was created for development sake. It is neither a delivery for Milestone 1 nor needs to be evaluated. All the changes are available along with the new features in Milestone 2 as a delivery.
> Also, I'de remove `testing_guide.md` from the `dot-marketplace-v2` repo, at least from the branches of these 3 milestones, as the testing guide is in the `dot_marketplace_docker` repo.
> 
We've removed the old testing guide from all the Phase 3 branches.

> Finally, I'd update (or remove, as you prefer) the part **Future Plans** from the `README.md` of `dot-marketplace-v2`, as it's the same for each Milestone's branch (maybe just keep it in `master`?).

We have modified this as per Phase 3.

Loaki07 commented:
> # Milestone 2
> As for the previous milestone, both docker and without-docker setups were straightforward.
> 
> ## Backend comments
> Everything seems to work greatly, I've just two things to say about `votes`:
> 
> * ### `accept_jury_duty` issue
>   I've just noted an issue, but I'm not sure if it's on your end, so I'll ask here: often, when I try to cast a vote with a `Potential Juror` I get this error from the frontend:
>   ![Screenshot 2023-01-12 at 13 27 24](https://user-images.githubusercontent.com/22419450/212080260-99e489d0-7dff-43fa-bff5-082cd2ac6afc.png)
>   So I tried to reproduce it with PolkadotJS to know more about it, and it gave me the `JurySelectionPeriodElapsed` error. From the [source code](https://github.com/WowLabz/dot-marketplace-v2/blob/bbeb45c2dfeeb91d2a4e39d191658361375992ef/pallets/pallet-tasking/src/lib.rs#L920), I see that it's like the `Juror` has been too slow to fire the `accept_jury_duty` function. Is that possible, and is that a bug?

The decision to be part of the jury is 1 day (1 era) or 14_400 slots i.e. 24 hours. However, for the purpose of testing it has been set to 5 blocks [source code](https://github.com/WowLabz/dot-marketplace-v2/blob/Phase3_Milestone2/pallets/pallet-tasking/src/lib.rs#L2124-L2153). Once the case begins/accepted, the jurors have 2 eras (2 days) to evaluate the work done & cast their votes to either of the parties i.e. the customer or worker. However, for testing purposes the duration is for 10 blocks and only 2 Jurors can cast their votes. 
I believe the error would have occurred, when 2 final Jurors have completed voting and the 3rd Juror was casting the vote. 

> * ### `accept_jury_duty` ratings
>   (This is just a discussion to understand better the project's operations)
>   I didn't get it why a `Juror` can, as well as choosing who they are voting for, cast their ratings for both the `customer` and `worker`. Is this to "correct" the ratings they're voting against? If yes, is it good for milestones disputes, as there are no ratings at the time a milestone is disputed? Or maybe is because a voter can express his rating in advance?
> 

Yes, the Jurors cast the vote for both `Custom` & `Worker` to ensure fair judgement on the milestone created & submitted as it would be reviewed completely at the time of dispute. The average of all the ratings provided will then be finalized. 

> 
> ## Frontend comments
> Other than this, I didn't see any other important things in the backend, but I'd like to comment quickly the frontend (I think it should be an MVP, but maybe these could still be useful to you):
> 
> * On the `login` page, I’d replace `User Name` with `Email` (and not writing under the input field “Enter registered EmailID”)

On the product page we have it in this manner. We'll fix this for the milestone 3.

> * On the `create project` page, if I try to change the wallet the frontend breaks, and then I have to logout and re-login;

We have found this issue, will fix it and let you know.

> * If I try `edit milestone`, its cost increases each time the button gets pressed;

We have fixed this issue. For explanation, the cost remained the same only that when viewing the cost the 2nd time it was showing in pico-units.

> * Isn't it possible to add an image and upload it in just one step instead of two?

Yes, we can fix this as well.

> * It looks a bit weird that just the `customer`, initially, must signup and login, but I guess it's for the MVP; by the way, what's the purpose of the signup?

Here, we are taking the Tags that the user is interested in, to show the Homepage to his desired Tags. This ecosystem also has few centralized components for User Management, Pagination, Tasks based on Interests,  etc.

> * Milestone's button is `Complete` for a `worker` who bidded also if the `customer` accepted a bid of another `worker` (if you then click it, it says "project does not exist").
> 

Could you please provide more information on this a screenshot may be.

> 
> ## Conclusion
> To accept the milestone, i'd like to get answers for the **Backend** part, but the **Frontend** one is definetly not urgent, as those are just comments for improvement / bugs findings.

Thank you reviewing in detail. The suggestions were helpful.

0xCaso commented:
@Loaki07 thanks a lot for the answers, and happy to see the feedback was useful. I've asked the team to accept the **Milestone 1**;

For the **Milestone 2**, I'm happy to accept also that one (as the comments you posted explain well the backend functioning) after the procedure for the first one is completed. In the meantime I can send you the screenshot for that frontend (minor) issue, and if you want you can also fix the other minor issues (it's an invite, not mandatory).

I'm also beginning the evaluation of **Milestone 3**.

Noc2 reviewed (changes requested):
@Loaki07 FYI, since milestone one is now approved. See https://github.com/w3f/Grant-Milestone-Delivery/pull/682. We could already pay you for the first milestone if you split this PR into three separate PRs.

0xCaso commented:
@Loaki07 consider, by the way, that I plan to accept everything before the end of today (if the Milestone 3 is good). So probably would be more convenient to keep just one PR @Noc2?

0xCaso commented:
> > * Milestone's button is `Complete` for a `worker` who bidded also if the `customer` accepted a bid of another `worker` (if you then click it, it says "project does not exist").
> 
> Could you please provide more information on this a screenshot may be.

@Loaki07 I've updated the Milestone 2 comment with two screenshots.

Loaki07 commented:
@0xCaso Thank you for the reviews. 

A quick update, the issue where the UI breaks when you switch wallet in-between a process has been fixed. Now, any time it occurs it will take you back to the homepage where you will need to connect your wallet again to proceed. These changes are pushed both for Milestone 2 & 3.

0xCaso reviewed (commented):
# Milestone 3

For this one, I've got just two questions:
- The `search_milestones` functionality works well, but is there a way to make it free or users must pay fees to perform searches?
- I couldn't upload the milestone's image (Click `Add File` -> I select the file -> Click `Upload`); could that be an issue with the NFT Storage service?
  
  ![Screenshot 2023-01-13 at 11 41 28](https://user-images.githubusercontent.com/22419450/212303774-e48fef64-b370-45b2-a663-8229353127d8.png)

Loaki07 commented:
> > > * Milestone's button is `Complete` for a `worker` who bidded also if the `customer` accepted a bid of another `worker` (if you then click it, it says "project does not exist").
> > 
> > 
> > Could you please provide more information on this a screenshot may be.
> 
> @Loaki07 I've updated the Milestone 2 comment with two screenshots.

Thank you for pointing this out, this issue is also fixed in Milestone 3.

Loaki07 commented:
> # Milestone 3
> For this one, I've got just two questions:
> 
> * The `search_milestones` functionality works well, but is there a way to make it free or users must pay fees to perform searches?

We've made the transaction fees as `Pays::No`. So, no money is charged for running the search.

> * I couldn't upload the milestone's image (Click `Add File` -> I select the file -> Click `Upload`); could that be an issue with the NFT Storage service?
>   ![Screenshot 2023-01-13 at 11 41 28](https://user-images.githubusercontent.com/22419450/212303774-e48fef64-b370-45b2-a663-8229353127d8.png)

The files are not uploading at all?

0xCaso commented:
> We've made the transaction fees as `Pays::No`. So, no money is charged for running the search.

Oh sorry, didn't see that. Great then!

> The files are not uploading at all?

I'm sorry, as I'm using a remote machine, I forgot to open the new ports `27019` and `8001`. Now everything works perfectly! Could you please add in the README of [`dot_marketplace_docker`](https://github.com/WowLabz/dot_marketplace_docker/tree/Phase3_Milestone3) which are the needed ports?

After that, I think everything else looks good.

Loaki07 commented:
> 

Great! We've added the ports to the README.md

0xCaso commented:
@Loaki07 All the milestones have been accepted, congratulations and thanks for the cooperation!

Just one last thing: recently we changed our guidelines, and I would ask you if it's possible to change USDT on Ethereum as payment method to another option choosable from [here](https://github.com/w3f/Grants-Program#pencil-process) (so if you want to stay with Ethereum you should change it to USDC or DAI).

If yes, it would be great if you could resend the three invoices changing the currency, and also adding the VAT Amount, which is 0.

Loaki07 commented:
Thank you! 😄 

Could you please confirm if Bitcoin payment is possible.

Loaki07 commented:
@0xCaso The payment address to Bitcoin is updated in this [pr](https://github.com/w3f/Grants-Program/pull/1439). Also, reuploaded the invoices for all the 3 milestones.

Loaki07 commented:
We are consulting a tax lawyer, we'll upload the correct invoices based  on the same.

Loaki07 commented:
@0xCaso 
Thank you for your patience. I've re-uploaded the 3 invoices. Kindly, check and let us know.

0xCaso commented:
Thanks again @Loaki07, I forwarded them internally! Please allow up to 14 days for the payment.

RouvenP commented:
hi @Loaki07 we transferred the payment last week

Loaki07 commented:
It's been received, thank you!
